### PR TITLE
Add `Iterator::array_chunks` (take N+1)

### PR DESCRIPTION
**Pull request by WaffleLapkin**
A revival of https://github.com/rust-lang/rust/pull/92393.

r? @Mark-Simulacrum 
cc @rossmacarthur @scottmcm @the8472 

I've tried to address most of the review comments on the previous attempt. The only thing I didn't address is `try_fold` implementation, I've left the "custom" one for now, not sure what exactly should it use.

### CODE REVIEW

**Comment by the8472** [comment]
Note that `iter.next_chunk()` exists now (#98326). And I expect these things to optimize poorly, so we'll need specialized implementations (as in #98553) for all sources that can yield chunks more efficiently and then plumb that through adapters that don't alter lengths like `Map`, `Copied`, `Take` etc. otherwise one won't get the autovectorized code one might expect.

I'm not sure if this the right approach and some design work is needed (#81615).

**Comment by scottmcm** [comment]
+1 that this should just use https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.next_chunk -- conveniently, that should mean that this PR gets *substantially* simpler, since the `unsafe` can live over there instead of here.

File: library/core/src/iter/adapters/array_chunks.rs
Comment (the8472) [inline library/core/src/iter/adapters/array_chunks.rs:122]: `by_ref` AND a double-reverse? 😬 that must be horrendously slow.

Imo we either need a `next_chunk_back` or only limit it to forward iteration. The former could be done in a separate PR, but we need to decide on an approach.

Comment (WaffleLapkin) [reply library/core/src/iter/adapters/array_chunks.rs:122]: I'm sorry you had to see this impl 😅

File: library/core/src/iter/adapters/array_chunks.rs
Comment (scottmcm) [inline library/core/src/iter/adapters/array_chunks.rs:114]: suggestion: You have a `Sized` iterator here (since it's a field), so you can avoid some of the `by_ref` optimization penalty by using https://github.com/rust-lang/rust/blob/master/library/core/src/iter/adapters/by_ref_sized.rs instead of `by_ref()`.

File: library/core/src/iter/traits/iterator.rs
Comment (scottmcm) [inline library/core/src/iter/traits/iterator.rs:3328]: Technically this doesn't panic like this.  I think it probably should -- at least for now, same as <https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.as_chunks> -- so please add an assert for that.

File: library/core/src/iter/adapters/array_chunks.rs
Comment (scottmcm) [inline library/core/src/iter/adapters/array_chunks.rs:90]: To avoid making a closure type generic over `I` too:
```suggestion
        self.try_fold(init, NeverShortCircuit::wrap_mut_2(f)).0
```
(and the same in `rfold`)

File: library/core/src/iter/traits/iterator.rs
Comment (scottmcm) [reply library/core/src/iter/traits/iterator.rs:3328]: Oh, nvm -- I see that it asserts in `new`.

**Review by scottmcm — CHANGES_REQUESTED** [review]
I have a few comments, but overall I think this is essentially fine for nightly.  (Lots of open questions for stabilization, but that's ok.)

Can you please open a tracking issue, add it to the unstable attributes, and address my other comments?

File: library/core/src/iter/adapters/array_chunks.rs
Comment (scottmcm) [inline library/core/src/iter/adapters/array_chunks.rs:157]: I don't think I'll block the PR on this, but `unwrap_err_unchecked` here scares me -- I added a note to the tracking issue.

My instinct for now is that `%N` is definitely trustworthy, so that combined with `take` is sufficient to say this is ok.  But I'm still afraid, since it's not `TrustedLen`, and wonder if someone can come up with an evil example here where it somehow gets a full chunk and thus is UB in safe code.

(At least the obvious things, like a `len()` that always says `usize`, is protected against because of the modulo.)

**Comment by scottmcm** [comment]
Running over this again I noticed one thing I'm wondering about, but I don't think it needs to block going to nightly.  Thanks for pushing this through!

@bors r+

File: library/core/src/iter/adapters/array_chunks.rs
Comment (the8472) [reply library/core/src/iter/adapters/array_chunks.rs:157]: As already noted in another comment I don't like this entire method anyway. Either we should add `next_chunk_back` or remove the DEI impl.

**Comment by the8472** [comment]
> (take N+1)

Clearly a bug! :D

File: library/core/src/iter/adapters/array_chunks.rs
Comment (WaffleLapkin) [reply library/core/src/iter/adapters/array_chunks.rs:157]: So, an evil impl can return less than `rem` elements, but `actual <= rem < N` still holds so the chunk will be an error no matter what.

An evil impl can just return a wrong length, so after this iterator returns a number of elements not divisible by `N`. But this is fine too, an `Err` will just be ignored:
https://github.com/rust-lang/rust/blob/5fbcde1b55c09421f43a7d6cfe09103d064ed2db/library/core/src/iter/adapters/array_chunks.rs#L116-L119

Comment (WaffleLapkin) [reply library/core/src/iter/adapters/array_chunks.rs:157]: That being said, as @the8472 already said, we should probably rewrite the DEI impl anyway (I'll try to do this, after this lands).

Comment (scottmcm) [reply library/core/src/iter/adapters/array_chunks.rs:157]: Yeah, the `%` is certainly fine.  What I'm not yet convinced of is that there's no way to make `take` misbehave somehow -- for example, `Take::try_fold` uses the inner `try_fold`, so maybe there'd be a way for that to be implemented wrongly-but-not-UB that could result in too many things getting put in the array somehow.

But I agree that just making a nicer implementation without the `unsafe{}` is the right plan.

Comment (WaffleLapkin) [reply library/core/src/iter/adapters/array_chunks.rs:157]: Actually you are right 😨 

Overriding `try_fold` with some nasty unsafe to skip over `Break(_)` allows you to trick `Take` into returning more elements that it should ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=223410f4b12798cd5490928290de878f)).

`n` can underflow, if a bad iterator impl skips over `Break(_)`:
https://github.com/rust-lang/rust/blob/2fbc08e2ce64dee45a29cb6133da6b32366268aa/library/core/src/iter/adapters/take.rs#L87-L89

The fact that unsafe code can't trust `Take` is scary.

Comment (scottmcm) [reply library/core/src/iter/adapters/array_chunks.rs:157]: I think the problem there is still in the `unsafe` in `Misbehave` -- it protects against double-drop, but duplicating arbitrary values can violate *safety* invariants in general.  (For example, if the type is `!Clone + !Default`, I can use you moving it into me to track resource consumption, like a ZST tracker for a global semaphore.)  So that specific example is just "well unsound `unsafe`-using code breaks everything".

But yeah, even though I've not been able to come up with a safe trick that would do it, things along those lines are what make me worried about it.  (And certainly if specialization existed then it would be possible to do particularly weird things in safe code because it can violate parametricity <sub>even worse than in normal code</sub>.)

Comment (WaffleLapkin) [reply library/core/src/iter/adapters/array_chunks.rs:157]: The example could check, that `type_id::<B>() == type_id::<usize>()` or something similar, which would IMO make the code sound.

Comment (scottmcm) [reply library/core/src/iter/adapters/array_chunks.rs:157]: Yeah, I thought about that one, but `TypeId::of::<B>()` needs `B: 'static`, which we don't have here.  Regardless, it's yet another chink in the armour that makes be scared we'll break through eventually.

**Pull request merged**
